SetupDataPkg/Library/ConfigVariableListLib/ConfigVariableListLib.c

```diff
@@ -15,9 +15,228 @@
+                                        VariableEntry. Updated on successful conversion and EFI_BUFFER_TOO_SMALL
+                                        returns.
+
+  @retval EFI_INVALID_PARAMETER   One or more input arguments are null.
+  @retval EFI_OUT_OF_RESOURCES    Memory allocation failed.
+  @retval EFI_BUFFER_TOO_SMALL    The input buffer does not contain a full variable list.
+  @retval EFI_COMPROMISED_DATA    The input variable list buffer has a corrupted CRC.
+  @retval EFI_SUCCESS             The operation succeeds.
+
+**/
+EFI_STATUS
+EFIAPI
+ConvertVariableEntryToVariableList (
+  IN  CONST CONFIG_VAR_LIST_ENTRY  *VariableEntry,
+  OUT VOID                         *VariableListBuffer,
+  IN OUT UINTN                     *Size
+  )
+{
+  EFI_STATUS  Status;
+  UINTN       NameSize;
+  UINTN       NeededSize;
+  UINTN       Offset;
+
+  // Sanity check for input parameters
+  if ((Size == NULL) || (VariableEntry == NULL) || ((VariableListBuffer == NULL) && (*Size != 0))) {
+    Status = EFI_INVALID_PARAMETER;
+    goto Exit;
+  }
+
+  // Sanity check 2 for input parameters
+  if ((VariableEntry->Name == NULL) || (VariableEntry->Data == NULL)) {
+    Status = EFI_INVALID_PARAMETER;
+    goto Exit;
+  }
+
+  NameSize = StrnSizeS (VariableEntry->Name, CONF_VAR_NAME_LEN);
+
+  NeededSize = VAR_LIST_SIZE (NameSize, VariableEntry->DataSize);
+
+  if (*Size < NeededSize) {
+    Status = EFI_BUFFER_TOO_SMALL;
+    *Size  = NeededSize;
+    goto Exit;
+  }
+
+  Offset = 0;
+  // Header
+  ((CONFIG_VAR_LIST_HDR *)VariableListBuffer)->NameSize = (UINT32)NameSize;
+  ((CONFIG_VAR_LIST_HDR *)VariableListBuffer)->DataSize = (UINT32)VariableEntry->DataSize;
+  Offset                                               += sizeof (CONFIG_VAR_LIST_HDR);
+
+  // Name
+  CopyMem ((UINT8 *)(VariableListBuffer) + Offset, VariableEntry->Name, NameSize);
+  Offset += NameSize;
+
+  // Guid
+  CopyMem ((UINT8 *)(VariableListBuffer) + Offset, &VariableEntry->Guid, sizeof (VariableEntry->Guid));
+  Offset += sizeof (VariableEntry->Guid);
+
+  // Attributes
+  CopyMem ((UINT8 *)(VariableListBuffer) + Offset, &VariableEntry->Attributes, sizeof (VariableEntry->Attributes));
+  Offset += sizeof (VariableEntry->Attributes);
+
+  // Data
+  CopyMem ((UINT8 *)(VariableListBuffer) + Offset, VariableEntry->Data, VariableEntry->DataSize);
+  Offset += VariableEntry->DataSize;
+
+  // CRC32
+  *(UINT32 *)((UINT8 *)(VariableListBuffer) + Offset) = CalculateCrc32 (VariableListBuffer, Offset);
+  Offset                                             += sizeof (UINT32);
+
+  *Size = NeededSize;
+
+  // They should still match in size...
+  ASSERT (Offset == NeededSize);
+
+  Status = EFI_SUCCESS;
+
+Exit:
+  return Status;
+}
+
 /**
   Parse Active Config Variable List and return full list or specific entry if VarName parameter != NULL
 
+  @param[out] VariableListBuffer  Pointer to raw variable list buffer.

```

Comment:
Should be [in] and VariableListBufferSize should be added to the comment.

Reply:
Agreed. Updated.